backend/models/sequence.js

```diff
@@ -19,16 +19,102 @@
 const db = require("../handler/db");
 const responseCodes = require("../response_codes.js");
 const utils = require("../utils");
+const yup = require("yup");
 
+const { update: updateGroup } = require("./group");
 const History = require("./history");
 const FileRef = require("./fileRef");
 const { getRefNodes } = require("./ref");
 const { getDefaultLegendId } = require("./modelSetting");
+const View = new (require("./view"))();
+const { cleanViewpoint, createViewpoint } = require("./viewpoint");
 
-const activityCol = (modelId) => `${modelId}.activities`;
 const sequenceCol = (modelId) => `${modelId}.sequences`;
 const legendCol = (modelId) => `${modelId}.sequences.legends`;
 
+const findMinMaxDate =  (frames) => {
+	let min = frames[0].dateTime;
+	let max = frames[0].dateTime;
+
+	frames.forEach((frame) => {
+		if (frame.dateTime < min) {
+			min = frame.dateTime;
+		}
+		if (frame.dateTime > max) {
+			max = frame.dateTime;
+		}
+	});
+
+	return {min, max};
+};
+
+const byDateTime = (frameA, frameB) => frameA.dateTime - frameB.dateTime;
+
+const notEmpty = (obj) => Object.keys(obj).length > 0;
+
+// Viewpoint checks
+const viewpointSchema = yup.object().test((viewpoint) => {
+	if (!viewpoint) {
+		return true;
+	}
+
+	const {transformation_group_ids, transformation_groups } = viewpoint;
+
+	// The viewpoints for sequences CANT have transformations groups
+	if (transformation_group_ids || transformation_groups) {
+		return false;
+	}
+
+	return true;
+});
+
+const frameSchema = yup.object().shape({
+	dateTime: yup.number().required(),
+	viewpoint: viewpointSchema,
+	viewId: yup.string()
+}).noUnknown().test(({viewId, viewpoint}) => {
+	// The frame must have either viewpoint or viewId, not both
+	return (viewId && !viewpoint) || (!viewId && viewpoint);
+});
+
+const nameSchema = yup.string().min(1).max(30);
+
+const sequenceSchema = yup.object().shape({
+	name: nameSchema.required(),
+	startDate: yup.number(),
+	endDate: yup.number(),
+	rev_id: yup.string(),
+	frames: yup.array().of(frameSchema).min(1).required()
+}).noUnknown().test(({startDate, endDate, frames}) => {
+	if ((startDate !== undefined || endDate !== undefined) && frames) {
+		const {min, max} = findMinMaxDate(frames);
+
+		// If the min date found in frames is smaller than the passed start date
+		// then the object is inconsistent
+		if(startDate !== undefined && min > startDate) {
+			return false;
+		}
+
+		// If the max date found in frames is larger than the passed end date
+		// then the object is inconsistent
+		if(endDate !== undefined && max < endDate) {
+			return false;
+		}
+	}
+
+	if (startDate !== undefined &&  endDate !== undefined && startDate > endDate) {
+		return false;
+	}
+
+	return true;
+});
+
+const sequenceEditSchema = sequenceSchema.shape({
+	name: nameSchema,
+	rev_id: yup.string().nullable(),
+	frames: yup.array().of(frameSchema).min(1)
```

Comment:
if we let user put in the startDate endDate then we should let the user edit the start date and end date?

Also if the frames are updated we may want to update the startDate/endDate?